{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "Nartsevipynb",
      "provenance": [],
      "collapsed_sections": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "fGRnVGY0I0t7",
        "outputId": "f2e1d25c-e25f-430f-938d-50dd08bbfead"
      },
      "source": [
        "!pip3 install biopython==1.77"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Collecting biopython==1.77\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/74/68/8d7bb5782a2531138c03f223e23bcec9be980a4a9be30e06351200220caf/biopython-1.77-cp37-cp37m-manylinux1_x86_64.whl (2.3MB)\n",
            "\u001b[K     |████████████████████████████████| 2.3MB 5.1MB/s \n",
            "\u001b[?25hRequirement already satisfied: numpy in /usr/local/lib/python3.7/dist-packages (from biopython==1.77) (1.19.5)\n",
            "Installing collected packages: biopython\n",
            "Successfully installed biopython-1.77\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "cQ7FPVsGIHcW"
      },
      "source": [
        "from Bio import SeqIO\n",
        "from Bio import Entrez\n",
        "from Bio.Seq import Seq\n",
        "from Bio.GenBank.Record import Record\n",
        "from Bio.SeqRecord import SeqRecord\n",
        "from Bio.Alphabet import IUPAC\n",
        "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
        "from Bio.Alphabet import generic_dna\n",
        "from collections import OrderedDict\n",
        "import pandas as pd\n",
        "import urllib\n",
        "import urllib.request\n",
        "import numpy as np\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "NSGMrvrLj3Sq"
      },
      "source": [
        "Задача: аннотировать фрагмент генома бактерии и как можно точнее определить таксон, к которому бактерия принадлежит. \n",
        "\n",
        "**Этапы работы**\n",
        "\n",
        "1) Получаем открытые рамки считывания (ОРС) через Artemis (Create->Mark open reading frames). Выделяем их все и смещаем к старт-кодонам (Edit->Trim selected features->To Any). Переименовываем их и задаем  Edit ->Automatically Create Gene Names (я задал my_orf с шагом 1)\n",
        "\n",
        "2) загружаем ORF из Artemis в blastp с параметрами:\n",
        "DataBase: uniprot/swiss-Prot\n",
        "Organism: Bacteria\n",
        "Max target sequences: 50\n",
        "Expect threshold: 0.5\n",
        "Word size: 6\n",
        "Matrix: BLOSUM62\n",
        "\n",
        "Ожидание составило минут 10. \n",
        "Далее выгружаем данные в формате HitTable.csv и начинаем анализировать\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "OZDiwKStc5lg"
      },
      "source": [
        "Ниже мы считываем данные из BLASP,а также результат работы artemis для координат CDS"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "IhVDDwdoJbT8",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 160
        },
        "outputId": "f12ada40-de73-45d9-9b56-7ccf8aa0edd6"
      },
      "source": [
        "#read all data info\n",
        "names=['query acc.ver',  'identity', 'alignment length','mismatches', 'gap opens', 'q_start', 'q_end','s. start', 's. end', 'evalue', 'bit score', '% positives']\n",
        "swiss = pd.read_csv('HitTableUprt.csv', names=names, index_col=0,)\n",
        "fasta_sequences = SeqIO.parse(open('DNA.fasta'),'fasta')\n",
        "with open('DNA.fasta') as out_file:\n",
        "    for fasta in fasta_sequences:\n",
        "        name = fasta.id \n",
        "        sequence = str(fasta.seq)\n",
        "coordinates = pd.DataFrame()\n",
        "with open('amino_acids.fasta', 'r') as text:\n",
        "  for line in text:\n",
        "    if 'my_orf' in line:\n",
        "      line = line.strip().split(' ')\n",
        "      cds = line[0][1:]\n",
        "      line = np.array(line[4].split(':')).astype(int)\n",
        "      coordinates[cds] = line\n",
        "     \n",
        "coordinates.index = ['begin', 'end']\n",
        "coordinates"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>my_orf001</th>\n",
              "      <th>my_orf001c</th>\n",
              "      <th>my_orf003</th>\n",
              "      <th>my_orf004</th>\n",
              "      <th>my_orf005</th>\n",
              "      <th>my_orf006c</th>\n",
              "      <th>my_orf007</th>\n",
              "      <th>my_orf008</th>\n",
              "      <th>my_orf009</th>\n",
              "      <th>my_orf002c</th>\n",
              "      <th>my_orf011</th>\n",
              "      <th>my_orf012c</th>\n",
              "      <th>my_orf013</th>\n",
              "      <th>my_orf014c</th>\n",
              "      <th>my_orf015</th>\n",
              "      <th>my_orf016</th>\n",
              "      <th>my_orf017</th>\n",
              "      <th>my_orf018</th>\n",
              "      <th>my_orf019c</th>\n",
              "      <th>my_orf003c</th>\n",
              "      <th>my_orf021</th>\n",
              "      <th>my_orf004c</th>\n",
              "      <th>my_orf023</th>\n",
              "      <th>my_orf005c</th>\n",
              "      <th>my_orf025</th>\n",
              "      <th>my_orf026</th>\n",
              "      <th>my_orf027</th>\n",
              "      <th>my_orf029</th>\n",
              "      <th>my_orf030c</th>\n",
              "      <th>my_orf031c</th>\n",
              "      <th>my_orf032</th>\n",
              "      <th>my_orf033c</th>\n",
              "      <th>my_orf034</th>\n",
              "      <th>my_orf035c</th>\n",
              "      <th>my_orf036</th>\n",
              "      <th>my_orf037c</th>\n",
              "      <th>my_orf038c</th>\n",
              "      <th>my_orf039</th>\n",
              "      <th>my_orf040</th>\n",
              "      <th>my_orf041c</th>\n",
              "      <th>...</th>\n",
              "      <th>my_orf064</th>\n",
              "      <th>my_orf065c</th>\n",
              "      <th>my_orf066</th>\n",
              "      <th>my_orf067c</th>\n",
              "      <th>my_orf068</th>\n",
              "      <th>my_orf069c</th>\n",
              "      <th>my_orf009c</th>\n",
              "      <th>my_orf071</th>\n",
              "      <th>my_orf010c</th>\n",
              "      <th>my_orf073</th>\n",
              "      <th>my_orf074c</th>\n",
              "      <th>my_orf011c</th>\n",
              "      <th>my_orf076</th>\n",
              "      <th>my_orf077</th>\n",
              "      <th>my_orf078c</th>\n",
              "      <th>my_orf012</th>\n",
              "      <th>my_orf080</th>\n",
              "      <th>my_orf081</th>\n",
              "      <th>my_orf082</th>\n",
              "      <th>my_orf083</th>\n",
              "      <th>my_orf084</th>\n",
              "      <th>my_orf085</th>\n",
              "      <th>my_orf086c</th>\n",
              "      <th>my_orf087</th>\n",
              "      <th>my_orf088c</th>\n",
              "      <th>my_orf089</th>\n",
              "      <th>my_orf090</th>\n",
              "      <th>my_orf091</th>\n",
              "      <th>my_orf092c</th>\n",
              "      <th>my_orf093</th>\n",
              "      <th>my_orf094</th>\n",
              "      <th>my_orf095</th>\n",
              "      <th>my_orf096c</th>\n",
              "      <th>my_orf097</th>\n",
              "      <th>my_orf098</th>\n",
              "      <th>my_orf013c</th>\n",
              "      <th>my_orf100c</th>\n",
              "      <th>my_orf101c</th>\n",
              "      <th>my_orf102</th>\n",
              "      <th>my_orf103c</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>begin</th>\n",
              "      <td>21</td>\n",
              "      <td>368</td>\n",
              "      <td>1573</td>\n",
              "      <td>2373</td>\n",
              "      <td>3289</td>\n",
              "      <td>17112</td>\n",
              "      <td>26915</td>\n",
              "      <td>6348</td>\n",
              "      <td>7555</td>\n",
              "      <td>8033</td>\n",
              "      <td>8256</td>\n",
              "      <td>8426</td>\n",
              "      <td>8961</td>\n",
              "      <td>9398</td>\n",
              "      <td>10185</td>\n",
              "      <td>11006</td>\n",
              "      <td>11710</td>\n",
              "      <td>12718</td>\n",
              "      <td>12851</td>\n",
              "      <td>13249</td>\n",
              "      <td>13493</td>\n",
              "      <td>13843</td>\n",
              "      <td>14199</td>\n",
              "      <td>14366</td>\n",
              "      <td>15177</td>\n",
              "      <td>15891</td>\n",
              "      <td>16981</td>\n",
              "      <td>19179</td>\n",
              "      <td>19783</td>\n",
              "      <td>20480</td>\n",
              "      <td>20574</td>\n",
              "      <td>21294</td>\n",
              "      <td>22253</td>\n",
              "      <td>22825</td>\n",
              "      <td>23414</td>\n",
              "      <td>23629</td>\n",
              "      <td>24047</td>\n",
              "      <td>24747</td>\n",
              "      <td>26205</td>\n",
              "      <td>26695</td>\n",
              "      <td>...</td>\n",
              "      <td>41030</td>\n",
              "      <td>41622</td>\n",
              "      <td>41941</td>\n",
              "      <td>43069</td>\n",
              "      <td>43994</td>\n",
              "      <td>44265</td>\n",
              "      <td>44521</td>\n",
              "      <td>45038</td>\n",
              "      <td>45398</td>\n",
              "      <td>45618</td>\n",
              "      <td>46111</td>\n",
              "      <td>46874</td>\n",
              "      <td>47099</td>\n",
              "      <td>47458</td>\n",
              "      <td>48279</td>\n",
              "      <td>49417</td>\n",
              "      <td>49813</td>\n",
              "      <td>50487</td>\n",
              "      <td>50934</td>\n",
              "      <td>51595</td>\n",
              "      <td>52442</td>\n",
              "      <td>52801</td>\n",
              "      <td>54036</td>\n",
              "      <td>54454</td>\n",
              "      <td>55601</td>\n",
              "      <td>55997</td>\n",
              "      <td>56941</td>\n",
              "      <td>58066</td>\n",
              "      <td>58962</td>\n",
              "      <td>58981</td>\n",
              "      <td>59745</td>\n",
              "      <td>61253</td>\n",
              "      <td>62236</td>\n",
              "      <td>63136</td>\n",
              "      <td>64152</td>\n",
              "      <td>66053</td>\n",
              "      <td>67012</td>\n",
              "      <td>68384</td>\n",
              "      <td>68681</td>\n",
              "      <td>69465</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>end</th>\n",
              "      <td>1460</td>\n",
              "      <td>685</td>\n",
              "      <td>1911</td>\n",
              "      <td>3050</td>\n",
              "      <td>4437</td>\n",
              "      <td>17633</td>\n",
              "      <td>27238</td>\n",
              "      <td>6803</td>\n",
              "      <td>8232</td>\n",
              "      <td>8425</td>\n",
              "      <td>8807</td>\n",
              "      <td>8812</td>\n",
              "      <td>9794</td>\n",
              "      <td>9841</td>\n",
              "      <td>10991</td>\n",
              "      <td>11719</td>\n",
              "      <td>12699</td>\n",
              "      <td>13479</td>\n",
              "      <td>13156</td>\n",
              "      <td>13617</td>\n",
              "      <td>14197</td>\n",
              "      <td>14205</td>\n",
              "      <td>15161</td>\n",
              "      <td>14677</td>\n",
              "      <td>15824</td>\n",
              "      <td>16625</td>\n",
              "      <td>19152</td>\n",
              "      <td>19748</td>\n",
              "      <td>20295</td>\n",
              "      <td>21022</td>\n",
              "      <td>21116</td>\n",
              "      <td>22055</td>\n",
              "      <td>22744</td>\n",
              "      <td>23595</td>\n",
              "      <td>23653</td>\n",
              "      <td>23934</td>\n",
              "      <td>25522</td>\n",
              "      <td>25055</td>\n",
              "      <td>26654</td>\n",
              "      <td>27723</td>\n",
              "      <td>...</td>\n",
              "      <td>41920</td>\n",
              "      <td>41924</td>\n",
              "      <td>42885</td>\n",
              "      <td>43824</td>\n",
              "      <td>45019</td>\n",
              "      <td>44648</td>\n",
              "      <td>45057</td>\n",
              "      <td>45586</td>\n",
              "      <td>45736</td>\n",
              "      <td>46049</td>\n",
              "      <td>46917</td>\n",
              "      <td>47176</td>\n",
              "      <td>47566</td>\n",
              "      <td>48231</td>\n",
              "      <td>50222</td>\n",
              "      <td>49758</td>\n",
              "      <td>50151</td>\n",
              "      <td>50915</td>\n",
              "      <td>51572</td>\n",
              "      <td>51912</td>\n",
              "      <td>52804</td>\n",
              "      <td>54435</td>\n",
              "      <td>54416</td>\n",
              "      <td>55710</td>\n",
              "      <td>55897</td>\n",
              "      <td>56944</td>\n",
              "      <td>58050</td>\n",
              "      <td>58797</td>\n",
              "      <td>59606</td>\n",
              "      <td>59487</td>\n",
              "      <td>60911</td>\n",
              "      <td>63022</td>\n",
              "      <td>62550</td>\n",
              "      <td>63648</td>\n",
              "      <td>66788</td>\n",
              "      <td>66364</td>\n",
              "      <td>68406</td>\n",
              "      <td>68698</td>\n",
              "      <td>69898</td>\n",
              "      <td>69782</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>2 rows × 101 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "       my_orf001  my_orf001c  my_orf003  ...  my_orf101c  my_orf102  my_orf103c\n",
              "begin         21         368       1573  ...       68384      68681       69465\n",
              "end         1460         685       1911  ...       68698      69898       69782\n",
              "\n",
              "[2 rows x 101 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 3
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "epKVGGEidD-d"
      },
      "source": [
        "Ниже - фильтр для работы BLASTP. Для него мы задаем порог identity 35% и e-value <0.001"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "HsT3sLJROGLb"
      },
      "source": [
        "def good_cds_selection(data_blast): \n",
        "  good = pd.DataFrame(columns = data_blast.columns)\n",
        "  for cds in data_blast.index:\n",
        "    curr_cds = pd.DataFrame(data_blast.loc[cds])\n",
        "    if curr_cds.shape[1] == 1:\n",
        "      curr_cds = curr_cds.T\n",
        "    if ((curr_cds.loc[:,'identity'].iloc[0] > 35) & (curr_cds.loc[:,'evalue'].iloc[0] < 0.001)):\n",
        "        good.loc[cds] = list(curr_cds.iloc[0])\n",
        "  return good\n",
        "good_cds = good_cds_selection(swiss)\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "xGrRwfkZ5Far",
        "outputId": "26952f7d-6cb9-40c4-8064-b10bdb280fdb"
      },
      "source": [
        "good_cds"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>query acc.ver</th>\n",
              "      <th>identity</th>\n",
              "      <th>alignment length</th>\n",
              "      <th>mismatches</th>\n",
              "      <th>gap opens</th>\n",
              "      <th>q_start</th>\n",
              "      <th>q_end</th>\n",
              "      <th>s. start</th>\n",
              "      <th>s. end</th>\n",
              "      <th>evalue</th>\n",
              "      <th>bit score</th>\n",
              "      <th>% positives</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>my_orf005</th>\n",
              "      <td>P96608.1</td>\n",
              "      <td>53.439</td>\n",
              "      <td>378</td>\n",
              "      <td>176</td>\n",
              "      <td>0</td>\n",
              "      <td>5</td>\n",
              "      <td>382</td>\n",
              "      <td>4</td>\n",
              "      <td>381</td>\n",
              "      <td>1.990000e-147</td>\n",
              "      <td>424.0</td>\n",
              "      <td>70.63</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>my_orf007</th>\n",
              "      <td>P96683.1</td>\n",
              "      <td>44.643</td>\n",
              "      <td>224</td>\n",
              "      <td>124</td>\n",
              "      <td>0</td>\n",
              "      <td>5</td>\n",
              "      <td>228</td>\n",
              "      <td>3</td>\n",
              "      <td>226</td>\n",
              "      <td>3.200000e-64</td>\n",
              "      <td>201.0</td>\n",
              "      <td>68.30</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>my_orf009</th>\n",
              "      <td>B5F9Z0.1</td>\n",
              "      <td>38.462</td>\n",
              "      <td>208</td>\n",
              "      <td>121</td>\n",
              "      <td>4</td>\n",
              "      <td>5</td>\n",
              "      <td>206</td>\n",
              "      <td>5</td>\n",
              "      <td>211</td>\n",
              "      <td>2.070000e-43</td>\n",
              "      <td>148.0</td>\n",
              "      <td>58.65</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>my_orf011</th>\n",
              "      <td>P41027.1</td>\n",
              "      <td>52.795</td>\n",
              "      <td>161</td>\n",
              "      <td>76</td>\n",
              "      <td>0</td>\n",
              "      <td>19</td>\n",
              "      <td>179</td>\n",
              "      <td>17</td>\n",
              "      <td>177</td>\n",
              "      <td>1.440000e-57</td>\n",
              "      <td>180.0</td>\n",
              "      <td>72.05</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>my_orf013</th>\n",
              "      <td>O07541.1</td>\n",
              "      <td>40.000</td>\n",
              "      <td>120</td>\n",
              "      <td>70</td>\n",
              "      <td>1</td>\n",
              "      <td>156</td>\n",
              "      <td>275</td>\n",
              "      <td>117</td>\n",
              "      <td>234</td>\n",
              "      <td>1.850000e-17</td>\n",
              "      <td>83.2</td>\n",
              "      <td>55.83</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>my_orf015</th>\n",
              "      <td>P42968.4</td>\n",
              "      <td>52.653</td>\n",
              "      <td>245</td>\n",
              "      <td>116</td>\n",
              "      <td>0</td>\n",
              "      <td>22</td>\n",
              "      <td>266</td>\n",
              "      <td>5</td>\n",
              "      <td>249</td>\n",
              "      <td>2.630000e-90</td>\n",
              "      <td>270.0</td>\n",
              "      <td>71.84</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>my_orf016</th>\n",
              "      <td>P60495.1</td>\n",
              "      <td>47.111</td>\n",
              "      <td>225</td>\n",
              "      <td>114</td>\n",
              "      <td>3</td>\n",
              "      <td>6</td>\n",
              "      <td>227</td>\n",
              "      <td>10</td>\n",
              "      <td>232</td>\n",
              "      <td>9.450000e-74</td>\n",
              "      <td>226.0</td>\n",
              "      <td>69.78</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>my_orf017</th>\n",
              "      <td>Q7WY77.2</td>\n",
              "      <td>48.624</td>\n",
              "      <td>327</td>\n",
              "      <td>160</td>\n",
              "      <td>4</td>\n",
              "      <td>3</td>\n",
              "      <td>322</td>\n",
              "      <td>1</td>\n",
              "      <td>326</td>\n",
              "      <td>2.020000e-105</td>\n",
              "      <td>313.0</td>\n",
              "      <td>67.89</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>my_orf018</th>\n",
              "      <td>Q81BT0.1</td>\n",
              "      <td>89.723</td>\n",
              "      <td>253</td>\n",
              "      <td>26</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>253</td>\n",
              "      <td>1</td>\n",
              "      <td>253</td>\n",
              "      <td>8.910000e-173</td>\n",
              "      <td>478.0</td>\n",
              "      <td>94.07</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>my_orf025</th>\n",
              "      <td>B7JDC3.1</td>\n",
              "      <td>93.953</td>\n",
              "      <td>215</td>\n",
              "      <td>13</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>215</td>\n",
              "      <td>1</td>\n",
              "      <td>215</td>\n",
              "      <td>2.980000e-150</td>\n",
              "      <td>417.0</td>\n",
              "      <td>97.67</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>my_orf026</th>\n",
              "      <td>Q81BT4.1</td>\n",
              "      <td>94.694</td>\n",
              "      <td>245</td>\n",
              "      <td>12</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>244</td>\n",
              "      <td>1</td>\n",
              "      <td>245</td>\n",
              "      <td>4.310000e-176</td>\n",
              "      <td>485.0</td>\n",
              "      <td>97.14</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>my_orf027</th>\n",
              "      <td>P54488.1</td>\n",
              "      <td>49.648</td>\n",
              "      <td>711</td>\n",
              "      <td>329</td>\n",
              "      <td>13</td>\n",
              "      <td>7</td>\n",
              "      <td>708</td>\n",
              "      <td>1</td>\n",
              "      <td>691</td>\n",
              "      <td>0.000000e+00</td>\n",
              "      <td>675.0</td>\n",
              "      <td>68.78</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>my_orf029</th>\n",
              "      <td>P37943.1</td>\n",
              "      <td>41.071</td>\n",
              "      <td>168</td>\n",
              "      <td>93</td>\n",
              "      <td>2</td>\n",
              "      <td>25</td>\n",
              "      <td>186</td>\n",
              "      <td>16</td>\n",
              "      <td>183</td>\n",
              "      <td>1.890000e-39</td>\n",
              "      <td>135.0</td>\n",
              "      <td>63.10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>my_orf030c</th>\n",
              "      <td>A1KMV3.1</td>\n",
              "      <td>39.535</td>\n",
              "      <td>172</td>\n",
              "      <td>97</td>\n",
              "      <td>5</td>\n",
              "      <td>2</td>\n",
              "      <td>166</td>\n",
              "      <td>71</td>\n",
              "      <td>242</td>\n",
              "      <td>1.430000e-31</td>\n",
              "      <td>115.0</td>\n",
              "      <td>61.63</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>my_orf034</th>\n",
              "      <td>P42301.1</td>\n",
              "      <td>40.000</td>\n",
              "      <td>165</td>\n",
              "      <td>88</td>\n",
              "      <td>3</td>\n",
              "      <td>1</td>\n",
              "      <td>158</td>\n",
              "      <td>1</td>\n",
              "      <td>161</td>\n",
              "      <td>5.930000e-39</td>\n",
              "      <td>132.0</td>\n",
              "      <td>56.36</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>my_orf035c</th>\n",
              "      <td>O07541.1</td>\n",
              "      <td>37.097</td>\n",
              "      <td>248</td>\n",
              "      <td>135</td>\n",
              "      <td>5</td>\n",
              "      <td>7</td>\n",
              "      <td>248</td>\n",
              "      <td>4</td>\n",
              "      <td>236</td>\n",
              "      <td>2.190000e-40</td>\n",
              "      <td>144.0</td>\n",
              "      <td>54.03</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>my_orf038c</th>\n",
              "      <td>P25737.5</td>\n",
              "      <td>62.994</td>\n",
              "      <td>481</td>\n",
              "      <td>173</td>\n",
              "      <td>3</td>\n",
              "      <td>13</td>\n",
              "      <td>489</td>\n",
              "      <td>3</td>\n",
              "      <td>482</td>\n",
              "      <td>0.000000e+00</td>\n",
              "      <td>632.0</td>\n",
              "      <td>79.42</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>my_orf041c</th>\n",
              "      <td>O34430.1</td>\n",
              "      <td>62.021</td>\n",
              "      <td>287</td>\n",
              "      <td>106</td>\n",
              "      <td>2</td>\n",
              "      <td>43</td>\n",
              "      <td>326</td>\n",
              "      <td>1</td>\n",
              "      <td>287</td>\n",
              "      <td>1.090000e-120</td>\n",
              "      <td>351.0</td>\n",
              "      <td>77.35</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>my_orf050</th>\n",
              "      <td>O31431.2</td>\n",
              "      <td>60.559</td>\n",
              "      <td>322</td>\n",
              "      <td>127</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>322</td>\n",
              "      <td>1</td>\n",
              "      <td>322</td>\n",
              "      <td>2.170000e-152</td>\n",
              "      <td>432.0</td>\n",
              "      <td>81.06</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>my_orf051</th>\n",
              "      <td>O31432.1</td>\n",
              "      <td>73.543</td>\n",
              "      <td>223</td>\n",
              "      <td>59</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>223</td>\n",
              "      <td>1</td>\n",
              "      <td>223</td>\n",
              "      <td>3.850000e-119</td>\n",
              "      <td>340.0</td>\n",
              "      <td>86.10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>my_orf052</th>\n",
              "      <td>O31433.1</td>\n",
              "      <td>63.208</td>\n",
              "      <td>318</td>\n",
              "      <td>116</td>\n",
              "      <td>1</td>\n",
              "      <td>9</td>\n",
              "      <td>326</td>\n",
              "      <td>1</td>\n",
              "      <td>317</td>\n",
              "      <td>4.200000e-139</td>\n",
              "      <td>398.0</td>\n",
              "      <td>81.13</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>my_orf053</th>\n",
              "      <td>P96728.2</td>\n",
              "      <td>65.000</td>\n",
              "      <td>100</td>\n",
              "      <td>29</td>\n",
              "      <td>2</td>\n",
              "      <td>12</td>\n",
              "      <td>110</td>\n",
              "      <td>4</td>\n",
              "      <td>98</td>\n",
              "      <td>5.480000e-39</td>\n",
              "      <td>127.0</td>\n",
              "      <td>76.00</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>my_orf054c</th>\n",
              "      <td>Q639P8.1</td>\n",
              "      <td>97.945</td>\n",
              "      <td>146</td>\n",
              "      <td>3</td>\n",
              "      <td>0</td>\n",
              "      <td>21</td>\n",
              "      <td>166</td>\n",
              "      <td>1</td>\n",
              "      <td>146</td>\n",
              "      <td>2.300000e-104</td>\n",
              "      <td>297.0</td>\n",
              "      <td>98.63</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>my_orf057c</th>\n",
              "      <td>P96681.1</td>\n",
              "      <td>37.554</td>\n",
              "      <td>466</td>\n",
              "      <td>283</td>\n",
              "      <td>8</td>\n",
              "      <td>32</td>\n",
              "      <td>493</td>\n",
              "      <td>11</td>\n",
              "      <td>472</td>\n",
              "      <td>1.260000e-101</td>\n",
              "      <td>315.0</td>\n",
              "      <td>61.37</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>my_orf063</th>\n",
              "      <td>P54562.1</td>\n",
              "      <td>49.359</td>\n",
              "      <td>156</td>\n",
              "      <td>79</td>\n",
              "      <td>0</td>\n",
              "      <td>8</td>\n",
              "      <td>163</td>\n",
              "      <td>1</td>\n",
              "      <td>156</td>\n",
              "      <td>1.050000e-51</td>\n",
              "      <td>164.0</td>\n",
              "      <td>62.82</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>my_orf064</th>\n",
              "      <td>Q52620.1</td>\n",
              "      <td>35.165</td>\n",
              "      <td>91</td>\n",
              "      <td>58</td>\n",
              "      <td>1</td>\n",
              "      <td>22</td>\n",
              "      <td>112</td>\n",
              "      <td>16</td>\n",
              "      <td>105</td>\n",
              "      <td>2.000000e-14</td>\n",
              "      <td>70.5</td>\n",
              "      <td>60.44</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>my_orf074c</th>\n",
              "      <td>P78271.1</td>\n",
              "      <td>50.725</td>\n",
              "      <td>69</td>\n",
              "      <td>33</td>\n",
              "      <td>1</td>\n",
              "      <td>6</td>\n",
              "      <td>73</td>\n",
              "      <td>1</td>\n",
              "      <td>69</td>\n",
              "      <td>5.540000e-13</td>\n",
              "      <td>68.9</td>\n",
              "      <td>66.67</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>my_orf078c</th>\n",
              "      <td>Q46306.1</td>\n",
              "      <td>45.776</td>\n",
              "      <td>651</td>\n",
              "      <td>339</td>\n",
              "      <td>6</td>\n",
              "      <td>4</td>\n",
              "      <td>646</td>\n",
              "      <td>5</td>\n",
              "      <td>649</td>\n",
              "      <td>0.000000e+00</td>\n",
              "      <td>584.0</td>\n",
              "      <td>66.36</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>my_orf081</th>\n",
              "      <td>P96708.1</td>\n",
              "      <td>53.957</td>\n",
              "      <td>139</td>\n",
              "      <td>64</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>139</td>\n",
              "      <td>21</td>\n",
              "      <td>159</td>\n",
              "      <td>7.580000e-47</td>\n",
              "      <td>151.0</td>\n",
              "      <td>73.38</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>my_orf082</th>\n",
              "      <td>P96707.1</td>\n",
              "      <td>70.192</td>\n",
              "      <td>208</td>\n",
              "      <td>62</td>\n",
              "      <td>0</td>\n",
              "      <td>5</td>\n",
              "      <td>212</td>\n",
              "      <td>2</td>\n",
              "      <td>209</td>\n",
              "      <td>3.300000e-106</td>\n",
              "      <td>306.0</td>\n",
              "      <td>81.73</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>my_orf083</th>\n",
              "      <td>P94425.1</td>\n",
              "      <td>45.455</td>\n",
              "      <td>88</td>\n",
              "      <td>48</td>\n",
              "      <td>0</td>\n",
              "      <td>9</td>\n",
              "      <td>96</td>\n",
              "      <td>1</td>\n",
              "      <td>88</td>\n",
              "      <td>1.880000e-25</td>\n",
              "      <td>93.2</td>\n",
              "      <td>65.91</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>my_orf085</th>\n",
              "      <td>Q813X6.1</td>\n",
              "      <td>72.558</td>\n",
              "      <td>215</td>\n",
              "      <td>59</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>215</td>\n",
              "      <td>1</td>\n",
              "      <td>215</td>\n",
              "      <td>6.340000e-105</td>\n",
              "      <td>327.0</td>\n",
              "      <td>84.19</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>my_orf089</th>\n",
              "      <td>P94474.2</td>\n",
              "      <td>46.557</td>\n",
              "      <td>305</td>\n",
              "      <td>162</td>\n",
              "      <td>1</td>\n",
              "      <td>9</td>\n",
              "      <td>312</td>\n",
              "      <td>11</td>\n",
              "      <td>315</td>\n",
              "      <td>9.570000e-99</td>\n",
              "      <td>295.0</td>\n",
              "      <td>69.84</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>my_orf094</th>\n",
              "      <td>P15555.2</td>\n",
              "      <td>35.860</td>\n",
              "      <td>343</td>\n",
              "      <td>205</td>\n",
              "      <td>5</td>\n",
              "      <td>50</td>\n",
              "      <td>386</td>\n",
              "      <td>41</td>\n",
              "      <td>374</td>\n",
              "      <td>3.220000e-67</td>\n",
              "      <td>220.0</td>\n",
              "      <td>57.14</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>my_orf095</th>\n",
              "      <td>Q81BX6.1</td>\n",
              "      <td>90.153</td>\n",
              "      <td>589</td>\n",
              "      <td>58</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>589</td>\n",
              "      <td>1</td>\n",
              "      <td>589</td>\n",
              "      <td>0.000000e+00</td>\n",
              "      <td>1126.0</td>\n",
              "      <td>95.59</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>my_orf096c</th>\n",
              "      <td>Q39X01.1</td>\n",
              "      <td>39.062</td>\n",
              "      <td>64</td>\n",
              "      <td>34</td>\n",
              "      <td>3</td>\n",
              "      <td>14</td>\n",
              "      <td>73</td>\n",
              "      <td>22</td>\n",
              "      <td>84</td>\n",
              "      <td>5.950000e-04</td>\n",
              "      <td>38.9</td>\n",
              "      <td>54.69</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>my_orf098</th>\n",
              "      <td>Q899Y1.1</td>\n",
              "      <td>47.297</td>\n",
              "      <td>814</td>\n",
              "      <td>399</td>\n",
              "      <td>7</td>\n",
              "      <td>93</td>\n",
              "      <td>877</td>\n",
              "      <td>56</td>\n",
              "      <td>868</td>\n",
              "      <td>0.000000e+00</td>\n",
              "      <td>774.0</td>\n",
              "      <td>66.58</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>my_orf100c</th>\n",
              "      <td>P26901.5</td>\n",
              "      <td>40.047</td>\n",
              "      <td>427</td>\n",
              "      <td>227</td>\n",
              "      <td>4</td>\n",
              "      <td>35</td>\n",
              "      <td>432</td>\n",
              "      <td>19</td>\n",
              "      <td>445</td>\n",
              "      <td>4.860000e-101</td>\n",
              "      <td>312.0</td>\n",
              "      <td>57.38</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>my_orf102</th>\n",
              "      <td>Q8R7C1.1</td>\n",
              "      <td>38.421</td>\n",
              "      <td>380</td>\n",
              "      <td>227</td>\n",
              "      <td>6</td>\n",
              "      <td>8</td>\n",
              "      <td>386</td>\n",
              "      <td>5</td>\n",
              "      <td>378</td>\n",
              "      <td>1.130000e-89</td>\n",
              "      <td>278.0</td>\n",
              "      <td>60.00</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "           query acc.ver  identity  ... bit score % positives\n",
              "my_orf005       P96608.1    53.439  ...     424.0       70.63\n",
              "my_orf007       P96683.1    44.643  ...     201.0       68.30\n",
              "my_orf009       B5F9Z0.1    38.462  ...     148.0       58.65\n",
              "my_orf011       P41027.1    52.795  ...     180.0       72.05\n",
              "my_orf013       O07541.1    40.000  ...      83.2       55.83\n",
              "my_orf015       P42968.4    52.653  ...     270.0       71.84\n",
              "my_orf016       P60495.1    47.111  ...     226.0       69.78\n",
              "my_orf017       Q7WY77.2    48.624  ...     313.0       67.89\n",
              "my_orf018       Q81BT0.1    89.723  ...     478.0       94.07\n",
              "my_orf025       B7JDC3.1    93.953  ...     417.0       97.67\n",
              "my_orf026       Q81BT4.1    94.694  ...     485.0       97.14\n",
              "my_orf027       P54488.1    49.648  ...     675.0       68.78\n",
              "my_orf029       P37943.1    41.071  ...     135.0       63.10\n",
              "my_orf030c      A1KMV3.1    39.535  ...     115.0       61.63\n",
              "my_orf034       P42301.1    40.000  ...     132.0       56.36\n",
              "my_orf035c      O07541.1    37.097  ...     144.0       54.03\n",
              "my_orf038c      P25737.5    62.994  ...     632.0       79.42\n",
              "my_orf041c      O34430.1    62.021  ...     351.0       77.35\n",
              "my_orf050       O31431.2    60.559  ...     432.0       81.06\n",
              "my_orf051       O31432.1    73.543  ...     340.0       86.10\n",
              "my_orf052       O31433.1    63.208  ...     398.0       81.13\n",
              "my_orf053       P96728.2    65.000  ...     127.0       76.00\n",
              "my_orf054c      Q639P8.1    97.945  ...     297.0       98.63\n",
              "my_orf057c      P96681.1    37.554  ...     315.0       61.37\n",
              "my_orf063       P54562.1    49.359  ...     164.0       62.82\n",
              "my_orf064       Q52620.1    35.165  ...      70.5       60.44\n",
              "my_orf074c      P78271.1    50.725  ...      68.9       66.67\n",
              "my_orf078c      Q46306.1    45.776  ...     584.0       66.36\n",
              "my_orf081       P96708.1    53.957  ...     151.0       73.38\n",
              "my_orf082       P96707.1    70.192  ...     306.0       81.73\n",
              "my_orf083       P94425.1    45.455  ...      93.2       65.91\n",
              "my_orf085       Q813X6.1    72.558  ...     327.0       84.19\n",
              "my_orf089       P94474.2    46.557  ...     295.0       69.84\n",
              "my_orf094       P15555.2    35.860  ...     220.0       57.14\n",
              "my_orf095       Q81BX6.1    90.153  ...    1126.0       95.59\n",
              "my_orf096c      Q39X01.1    39.062  ...      38.9       54.69\n",
              "my_orf098       Q899Y1.1    47.297  ...     774.0       66.58\n",
              "my_orf100c      P26901.5    40.047  ...     312.0       57.38\n",
              "my_orf102       Q8R7C1.1    38.421  ...     278.0       60.00\n",
              "\n",
              "[39 rows x 12 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 6
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "0_i057ytdUow"
      },
      "source": [
        "Для всех хороших находок удаляем пересекающиеся с ними CDS с приоритетом по identity. Также необходимо было учесть tRNA из tRNA-scan и/или Aragorn, но оба тула ничего не нашли (файлы прилагаю)\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "7rEfS-sRKaHJ",
        "outputId": "aa5fa38e-c7ca-4859-81c5-a2e5771c70ad"
      },
      "source": [
        "def intersect(st1, end1, st2, end2):\n",
        "      r = ((st1 <= end2) & (end1 >= end2))\n",
        "      l = ((end1 >= st2) & (st1 <= st2))\n",
        "      return (r | l)\n",
        "\n",
        "print(intersect(5, 10, 7,8)) #must be true\n",
        "print(intersect(5, 10, 2,8)) #must be true\n",
        "print(intersect(5, 10, 8, 13)) #must be true\n",
        "print(intersect(5, 10, 100, 200))#must be false\n",
        "print('________________________________________')\n",
        "\n",
        "def parse_intersections(findings, coord): \n",
        "  all_cds, good = list(findings.index), []\n",
        "  findings = findings.sort_values(by='identity', ascending=False)\n",
        "  coord = coord.T.loc[findings.index]\n",
        "  while len(all_cds) != 2:\n",
        "     curr_cds = all_cds[0]\n",
        "     all_cds.remove(curr_cds)\n",
        "     good.append(curr_cds)\n",
        "     for cds in all_cds:\n",
        "       if intersect(coord.loc[curr_cds, 'begin'], coord.loc[curr_cds, 'end'],\n",
        "                    coord.loc[cds, 'begin'], coord.loc[cds, 'end']):\n",
        "         all_cds.remove(cds)\n",
        "  return good\n",
        "unitersected = parse_intersections(good_cds, coordinates)\n",
        "coordinates.T.loc[unitersected].shape"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "True\n",
            "True\n",
            "True\n",
            "False\n",
            "________________________________________\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "(34, 2)"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 8
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "USbcE3cadmLk"
      },
      "source": [
        "Участки между CDS заносим в отдельный файл для BLASTN и запоминаем для них координаты"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "-9xoqgul5TMk"
      },
      "source": [
        "agtc_fa, name_fa = [],  []\n",
        "good = coordinates.T.loc[unitersected]\n",
        "spaces_coord = pd.DataFrame()\n",
        "for cds in range(len(unitersected)-1):\n",
        "  agtc_fa.append(sequence[good.iloc[cds, 1]:good.iloc[cds+1, 0]])\n",
        "  name_fa.append('seq'+str(cds))\n",
        "  spaces_coord['seq'+str(cds)] = [coordinates.T.loc[unitersected].iloc[cds, 1], coordinates.T.loc[unitersected].iloc[cds+1, 0]]\n",
        "ofile = open(\"seq_nucl_spaces.fasta\", \"w\")\n",
        "spaces_coord = spaces_coord.T\n",
        "spaces_coord.columns = ['begin', 'end']\n",
        "for i in range(len(agtc_fa)):#\n",
        "      ofile.write(\">\" + name_fa[i] + \"\\n\" +agtc_fa[i] + \"\\n\")\n",
        "ofile.close()"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "bqIuPhX4dx6_"
      },
      "source": [
        "Результат для BLASTN пропускаем через фильтр по identity и e-value. Так как мы получаем выравнивание и не имеем информации о том, где на самом деле лежит начало важного учатска и его конец, то необходимо учесть координаты выравнивания (и следить за тем, чтобы ничего не пересеклось)"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "EFavA-8OXB2q"
      },
      "source": [
        "names=[' ', 'query acc.ver',  'identity', 'alignment length','mismatches', 'gap opens', 'q_start', 'q_end','s. start', 's. end', 'evalue', 'bit score']\n",
        "blastn = pd.read_csv('blastn.csv', index_col=0, names=names)\n",
        "blastn = good_cds_selection(blastn)\n",
        "blastn = blastn.drop('seq0') #много мисматчей и вообще вызывает подозрения\n",
        "for x in blastn.index:\n",
        "  blastn.loc[x, 'q_start'] += spaces_coord.loc[x, 'begin'] \n",
        "  blastn.loc[x, 'q_end'] += spaces_coord.loc[x, 'begin'] \n",
        "\n",
        "temp = blastn.loc[:,['q_start',\t'q_end']]\n",
        "temp.columns = ['begin', 'end']\n",
        "all_coordinates = pd.concat([coordinates.T.loc[unitersected], temp]).sort_values(by='begin')\n",
        "annotation_of_cds = pd.concat([blastn, good_cds.loc[unitersected]])\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "yPCCDY2Kr3Ye"
      },
      "source": [
        "для orf с низким identity попробуем запустить interPro (ниже указаны только те, для которых было что-то найдено, для остальных None predicted)"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "pgLjYjZ5piaB"
      },
      "source": [
        "intrtPro_result = {'my_orf064': 'Transcription regulator HTH, AraC- type (IPR020449)',\n",
        "                   'my_orf009': 'Copper homeostasis protein CutC (IPR005627)',\n",
        "                   'my_orf029': 'Peptidase S26A, signal peptidase I (IPR000223)',\n",
        "                   'my_orf011': 'Peptidase S26A, signal peptidase I (IPR000223)'}\n",
        "\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Dkc97s6vd76P"
      },
      "source": [
        "Создаем аннотацию"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Bajbvn6fH4gn",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "28e872bc-2647-4d4a-ab29-0b722aa9997a"
      },
      "source": [
        "sequence_object = Seq(sequence, generic_dna)\n",
        "record = SeqRecord(sequence_object, id=' ', name='Example',annotations={\"molecule_type\": \"DNA\", \"organism\": \"Bacillus subtilis\"},description='bacteria Bacillus subtilis')\n",
        "for cds in all_coordinates.index:\n",
        "  print(cds)\n",
        "  if 'c' in cds:\n",
        "        strand = -1\n",
        "  else:\n",
        "        strand = 1\n",
        "  info = (str(annotation_of_cds.loc[cds,'identity']) + ', ' +'e-value: ' + str(annotation_of_cds.loc[cds,'evalue']) + ', ' + '% positives: ' + str(annotation_of_cds.loc[cds,'% positives']))\n",
        "  if  cds[:3] == 'seq':\n",
        "    handle = Entrez.efetch(db=\"nucleotide\", id=annotation_of_cds.loc[cds, 'query acc.ver'], rettype=\"gb\", retmode=\"text\")\n",
        "    for seq_record in SeqIO.parse(handle, \"gb\"):\n",
        "      info = (seq_record.annotations['organism'] + \" \" + info)\n",
        "    qualifier=OrderedDict([('gene',[annotation_of_cds.loc[cds, 'query acc.ver']]),('product',[seq_record.description]),('note',[info])])\n",
        "    feature = SeqFeature(FeatureLocation(start=int(all_coordinates.loc[cds, 'begin']),end=int(all_coordinates.loc[cds, 'end']), strand=strand), type=\"CDS\",qualifiers=qualifier)\n",
        "    record.features.append(feature)\n",
        "  elif cds in intrtPro_result.keys():\n",
        "    with urllib.request.urlopen('http://www.uniprot.org/uniprot/'+ annotation_of_cds.loc[cds, 'query acc.ver'] + '.xml') as url:\n",
        "        link_info = SeqIO.read(url, \"uniprot-xml\")\n",
        "    qualifier=OrderedDict([('gene',[intrtPro_result[cds]]),('product',[link_info.description]),('note',[info+'interPRO result']),('translation', [link_info.seq])])\n",
        "    feature = SeqFeature(FeatureLocation(start=int(all_coordinates.loc[cds, 'begin']),end=int(all_coordinates.loc[cds, 'end']), strand=strand), type=\"CDS\",qualifiers=qualifier)\n",
        "    record.features.append(feature)\n",
        "    print('iPro')\n",
        "\n",
        "\n",
        "  else: \n",
        "      with urllib.request.urlopen('http://www.uniprot.org/uniprot/'+ annotation_of_cds.loc[cds, 'query acc.ver'] + '.xml') as url:\n",
        "        link_info = SeqIO.read(url, \"uniprot-xml\")\n",
        "      if 'organism' in link_info.annotations:\n",
        "        info += link_info.annotations['organism']\n",
        "      if 'gene_name_primary' in link_info.annotations:\n",
        "        qualifier=OrderedDict([('gene',[link_info.annotations['gene_name_primary']]),('product',[link_info.description]),('note',[info]),('translation', [link_info.seq])])\n",
        "      else: \n",
        "        qualifier=OrderedDict([('gene',[link_info.description]),('product',[link_info.description]),('note',[info]),('translation', [link_info.seq])])\n",
        "\n",
        "      feature = SeqFeature(FeatureLocation(start=int(all_coordinates.loc[cds, 'begin']),end=int(all_coordinates.loc[cds, 'end']), strand=strand), type=\"CDS\",qualifiers=qualifier)\n",
        "      record.features.append(feature)\n",
        " \n",
        "with open('Nartsev_annot.gb', 'w') as output_file:\n",
        "  SeqIO.write(record, output_file, 'genbank')"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "my_orf005\n",
            "my_orf009\n",
            "iPro\n",
            "my_orf011\n",
            "iPro\n",
            "seq3\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "stream",
          "text": [
            "/usr/local/lib/python3.7/dist-packages/Bio/Entrez/__init__.py:668: UserWarning: \n",
            "Email address is not specified.\n",
            "\n",
            "To make use of NCBI's E-utilities, NCBI requires you to specify your\n",
            "email address with each request.  As an example, if your email address\n",
            "is A.N.Other@example.com, you can specify it as follows:\n",
            "   from Bio import Entrez\n",
            "   Entrez.email = 'A.N.Other@example.com'\n",
            "In case of excessive usage of the E-utilities, NCBI will attempt to contact\n",
            "a user at the email address provided before blocking access to the\n",
            "E-utilities.\n",
            "  UserWarning,\n"
          ],
          "name": "stderr"
        },
        {
          "output_type": "stream",
          "text": [
            "my_orf013\n",
            "seq4\n",
            "my_orf015\n",
            "my_orf016\n",
            "seq6\n",
            "my_orf018\n",
            "seq7\n",
            "my_orf025\n",
            "seq8\n",
            "my_orf026\n",
            "seq9\n",
            "my_orf027\n",
            "my_orf029\n",
            "iPro\n",
            "seq11\n",
            "my_orf030c\n",
            "seq12\n",
            "my_orf034\n",
            "seq13\n",
            "my_orf035c\n",
            "seq14\n",
            "my_orf038c\n",
            "seq15\n",
            "my_orf041c\n",
            "my_orf007\n",
            "seq16\n",
            "my_orf050\n",
            "seq17\n",
            "my_orf051\n",
            "seq18\n",
            "my_orf053\n",
            "seq19\n",
            "my_orf054c\n",
            "seq20\n",
            "my_orf057c\n",
            "seq21\n",
            "my_orf063\n",
            "seq22\n",
            "my_orf064\n",
            "iPro\n",
            "seq23\n",
            "my_orf074c\n",
            "seq24\n",
            "my_orf078c\n",
            "seq25\n",
            "my_orf081\n",
            "my_orf082\n",
            "my_orf083\n",
            "seq28\n",
            "my_orf085\n",
            "seq29\n",
            "my_orf089\n",
            "seq30\n",
            "my_orf094\n",
            "seq31\n",
            "my_orf095\n",
            "seq32\n",
            "my_orf098\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "YX5LY9udeAEQ"
      },
      "source": [
        "Считываем информацию из RAST"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "VtiqBmIOKj4_"
      },
      "source": [
        "rast_cds = pd.DataFrame()\n",
        "with open ('rast.txt') as text:\n",
        "  for i, line in enumerate(text):\n",
        "    line = line.split('\t')[1][3:]\n",
        "    coord_gene = np.array(line.split('_')).astype(int)\n",
        "    if coord_gene[0] > coord_gene[1]:\n",
        "      coord_gene[0], coord_gene[1] = coord_gene[1], coord_gene[0]\n",
        "      rast_cds[i] = coord_gene[0], coord_gene[1]\n",
        "    else:\n",
        "      rast_cds[i] = coord_gene[0], coord_gene[1]\n",
        "\n",
        "rast_cds = rast_cds.T\n",
        "rast_cds.index = range(len(rast_cds.index))"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "yPX0tjwzBBMx",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 419
        },
        "outputId": "5ddbb02e-6b99-4409-ba61-8cf929346606"
      },
      "source": [
        "rast_cds"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>0</th>\n",
              "      <th>1</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>21</td>\n",
              "      <td>1460</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>1573</td>\n",
              "      <td>1911</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>1979</td>\n",
              "      <td>2092</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>2391</td>\n",
              "      <td>3050</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>3289</td>\n",
              "      <td>4437</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>73</th>\n",
              "      <td>61268</td>\n",
              "      <td>63022</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>74</th>\n",
              "      <td>63136</td>\n",
              "      <td>63648</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>75</th>\n",
              "      <td>64152</td>\n",
              "      <td>66788</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>76</th>\n",
              "      <td>67012</td>\n",
              "      <td>68391</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>77</th>\n",
              "      <td>68681</td>\n",
              "      <td>69898</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>78 rows × 2 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "        0      1\n",
              "0      21   1460\n",
              "1    1573   1911\n",
              "2    1979   2092\n",
              "3    2391   3050\n",
              "4    3289   4437\n",
              "..    ...    ...\n",
              "73  61268  63022\n",
              "74  63136  63648\n",
              "75  64152  66788\n",
              "76  67012  68391\n",
              "77  68681  69898\n",
              "\n",
              "[78 rows x 2 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 17
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Q6cOKYczeEqB"
      },
      "source": [
        "Вспомогательная функция для подсчета IoU (в числитель идет пересечение)"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "COyEdgQEIMPE",
        "outputId": "5ac320b7-ed23-4997-9b0d-c1b947e4ca7b"
      },
      "source": [
        "def covering(st1, end1, st2, end2):\n",
        "  if st1 > end1:\n",
        "    st1, end1 = end1, st1\n",
        "  if st2 > end2:\n",
        "    st2, end2 = end2, st2\n",
        "  if (st1 < st2) and (end2 > end1):\n",
        "    overlap = end1 - st2\n",
        "  elif (st2>st1) and (end1>end2):\n",
        "    overlap = end2-st2\n",
        "  elif (st2<st1) and (end1>end2):\n",
        "    overlap = end2 - st1\n",
        "  else:\n",
        "    overlap = end1 - st1\n",
        "  if overlap < 0:\n",
        "    return 0\n",
        "  else:\n",
        "    return overlap+1 \n",
        "\n",
        "covering(1,10,5,7) #for test\n",
        "covering(2,10,1,100)#for test\n",
        "covering(4,7,6,2)#for test\n",
        "covering(4,7,6,2)#for test\n",
        "covering(10,15,12,17)#for test\n",
        "covering(10, 100, 200,300)#for test\n",
        "covering(1, 10, 1, 10)#for test"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "10"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 14
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "OXIp7uRYeNAf"
      },
      "source": [
        "Считаем IoU (поделить сумму длин пересекающихся участков на сумму объединения по всем генам)"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "VcIFcORwLqWv"
      },
      "source": [
        "inter = []\n",
        "denominator = []\n",
        "s = 0\n",
        "for i in rast_cds.index:\n",
        "  maximum = 0\n",
        "  for k in range(len(all_coordinates.index)): \n",
        "    value = covering(rast_cds.iloc[i,0], rast_cds.iloc[i,1],\n",
        "                 all_coordinates.iloc[k,0], all_coordinates.iloc[k,1])\n",
        "    if value > maximum:\n",
        "      maximum = value\n",
        "      s = k\n",
        "  denominator.append( (rast_cds.iloc[i,1] - rast_cds.iloc[i,0]) + (all_coordinates.iloc[s,1] - all_coordinates.iloc[s,0]) - maximum)\n",
        "  inter.append( maximum)\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "B_NpJ6jSRkVs",
        "outputId": "ffba967a-13e2-41ec-ea52-1b31841aae44"
      },
      "source": [
        "iou = sum(inter)/sum(denominator)\n",
        "iou"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "0.40335619237930875"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 16
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "doQBg2E4GYgu"
      },
      "source": [
        "Посчитаем, сколько совпало позиций начала и конца:"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "kwMbPAumQ4au"
      },
      "source": [
        "st = 0\n",
        "for i in rast_cds.index:\n",
        "   for k in range(len(all_coordinates.index)):\n",
        "     if rast_cds.iloc[i,0] == all_coordinates.iloc[k,0]:\n",
        "       st+=1\n",
        "end = 0\n",
        "for i in rast_cds.index:\n",
        "   for k in range(len(all_coordinates.index)):\n",
        "     if rast_cds.iloc[i,1] == all_coordinates.iloc[k,1]:\n",
        "       end+=1\n",
        "\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "FS-n4WUYhWHw"
      },
      "source": [
        "st - сколько совпало начальных позиций\n",
        "end - сколько совпало конечных позиций"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "G6GyGF1mWXrb",
        "outputId": "7646714e-b8ba-4011-cf1f-c55464fb269d"
      },
      "source": [
        "st, end"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "(25, 28)"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 31
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "2hS20gDzp9mi"
      },
      "source": [
        "Ниже аналогичный рассчет для Prokka\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "yg7CP5t4WvBG"
      },
      "source": [
        "prokka = {}\n",
        "i=0\n",
        "with open ('Galaxy108-_Prokka_on_data_100_tbl (1).txt') as text:\n",
        "  for line in text:\n",
        "    if 'CDS' in line:\n",
        "      line = line.split('\t')\n",
        "      prokka[i] = [int(x) for x in line[:2]]\n",
        "      i+=1\n",
        "prokka = pd.DataFrame(prokka).T"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "R0L7MIIypw-V"
      },
      "source": [
        "inter = []\n",
        "denominator = []\n",
        "s = 0\n",
        "for i in prokka.index:\n",
        "  maximum = 0\n",
        "  for k in range(len(all_coordinates.index)): \n",
        "    value = covering(prokka.iloc[i,0], prokka.iloc[i,1],\n",
        "                 all_coordinates.iloc[k,0], all_coordinates.iloc[k,1])\n",
        "    if value > maximum:\n",
        "      maximum = value\n",
        "      s = k\n",
        "  denominator.append( (prokka.iloc[i,1] - prokka.iloc[i,0]) + (all_coordinates.iloc[s,1] - all_coordinates.iloc[s,0]) - maximum)\n",
        "  inter.append( maximum)\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "alqwKB7Ap31p",
        "outputId": "5180710a-e3d1-4b5c-d6a1-1614846426ba"
      },
      "source": [
        "iou = sum(inter)/sum(denominator)\n",
        "iou"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "0.5176267853144197"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 34
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "sFs7MW1rrrfO"
      },
      "source": [
        "посмотрим на количество совпадающих позиций"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "EHkeooMdp5db"
      },
      "source": [
        "st = 0\n",
        "for i in prokka.index:\n",
        "   for k in range(len(all_coordinates.index)):\n",
        "     if prokka.iloc[i,0] == all_coordinates.iloc[k,0]:\n",
        "       st+=1\n",
        "end = 0\n",
        "for i in prokka.index:\n",
        "   for k in range(len(all_coordinates.index)):\n",
        "     if prokka.iloc[i,1] == all_coordinates.iloc[k,1]:\n",
        "       end+=1\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "lPd1TWTSqH75",
        "outputId": "7342b133-d776-414b-b808-2a90c0ad1fc8"
      },
      "source": [
        "st, end"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "(17, 33)"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 36
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "_I4iZcxlhgHX"
      },
      "source": [
        "Вывод: \n",
        "Организм - Bacillus subtilis. Количество хороших находок в результате работы blast - 30. Еще 4 находки были уточнены с помощью InterPro. \n",
        "\n",
        "Наибольшее совпадение по по проаннотированным участкам с аннотацией с результатом работы Prokka"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "11ol-nClXOGP",
        "outputId": "d11d86ab-c46c-45e3-9a3f-f826dbbfc516"
      },
      "source": [
        "prokka_fuc = []\n",
        "with open ('Galaxy108-_Prokka_on_data_100_tbl (1).txt') as text:\n",
        "  for line in text:\n",
        "    if (('product' in line) and ('hypothetical' not in line)):\n",
        "      line = line.strip().split('\t')\n",
        "      prokka_fuc.append(line[1])\n",
        "print(len(prokka_fuc))\n",
        "prokka_fuc\n"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "33\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "['Putative acyl-CoA dehydrogenase YdbM',\n",
              " 'Copper homeostasis protein CutC',\n",
              " 'Signal peptidase IB',\n",
              " 'HTH-type transcriptional regulator KipR',\n",
              " '5-oxoprolinase subunit B',\n",
              " '5-oxoprolinase subunit C',\n",
              " '5-oxoprolinase subunit A',\n",
              " 'Pyrrolidone-carboxylate peptidase',\n",
              " 'NAD-dependent protein deacetylase',\n",
              " 'Penicillin-binding protein H',\n",
              " 'Signal peptidase I P',\n",
              " 'Rhamnosyl O-methyltransferase',\n",
              " '8-oxo-dGTP diphosphatase',\n",
              " 'putative protein YhaZ',\n",
              " 'Lysine-specific permease',\n",
              " 'Uridine kinase',\n",
              " 'Transcriptional activator protein CopR',\n",
              " 'Adaptive-response sensory-kinase SasA',\n",
              " 'HTH-type transcriptional regulator NorG',\n",
              " 'putative amino-acid metabolite efflux pump',\n",
              " 'putative protein YqjY',\n",
              " 'DNA gyrase inhibitor',\n",
              " 'RNA pyrophosphohydrolase',\n",
              " 'Tetracycline resistance protein TetO',\n",
              " 'Transcriptional regulator SlyA',\n",
              " 'Putative NAD(P)H nitroreductase YodC',\n",
              " 'Putative monooxygenase YcnE',\n",
              " 'ATPase RavA',\n",
              " 'IS4 family transposase MICBce1',\n",
              " 'Putative adenine deaminase YerA',\n",
              " 'Collagenase ColT',\n",
              " 'Vegetative catalase',\n",
              " 'Acetylornithine aminotransferase']"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 53
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "dnCfXLBeh7UQ",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "83d249ce-f286-482c-cd78-6fc0b7caba90"
      },
      "source": [
        "rast_fuc = []\n",
        "with open ('rast.txt') as text:\n",
        "  for i, line in enumerate(text):\n",
        "    print(line)\n",
        "    line = line.strip().split('\t')[2]\n",
        "    if 'hypothetical' not in line:\n",
        "     rast_fuc.append(line)\n",
        "print(len(set(set(rast_fuc))))\n",
        "set(rast_fuc)"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "fig|2116814.3352.peg.1\t14_21_1460\tPredicted histidine uptake transporter\n",
            "\n",
            "fig|2116814.3352.peg.2\t14_1573_1911\tacetyltransferase, GNAT family\n",
            "\n",
            "fig|2116814.3352.peg.3\t14_1979_2092\tacetyltransferase, GNAT family\n",
            "\n",
            "fig|2116814.3352.peg.4\t14_2391_3050\tEpiH/GdmH-related protein\n",
            "\n",
            "fig|2116814.3352.peg.5\t14_3289_4437\tButyryl-CoA dehydrogenase (EC 1.3.99.2)\n",
            "\n",
            "fig|2116814.3352.peg.6\t14_5138_4668\tFIG01226774: hypothetical protein\n",
            "\n",
            "fig|2116814.3352.peg.7\t14_5208_5915\tTranscriptional regulator, ArsR family\n",
            "\n",
            "fig|2116814.3352.peg.8\t14_6348_6803\tGlyoxalase family protein\n",
            "\n",
            "fig|2116814.3352.peg.9\t14_7555_8232\tCytoplasmic copper homeostasis protein cutC\n",
            "\n",
            "fig|2116814.3352.peg.10\t14_8256_8807\tSignal peptidase I (EC 3.4.21.89)\n",
            "\n",
            "fig|2116814.3352.peg.11\t14_8961_9794\tDNA alkylation repair enzyme\n",
            "\n",
            "fig|2116814.3352.peg.12\t14_10124_10237\thypothetical protein\n",
            "\n",
            "fig|2116814.3352.peg.13\t14_10239_10991\tTranscriptional regulator, IclR family\n",
            "\n",
            "fig|2116814.3352.peg.14\t14_11006_11719\tAllophanate hydrolase 2 subunit 1 (EC 3.5.1.54)\n",
            "\n",
            "fig|2116814.3352.peg.15\t14_11710_12699\tAllophanate hydrolase 2 subunit 2 (EC 3.5.1.54)\n",
            "\n",
            "fig|2116814.3352.peg.16\t14_12718_13479\tLactam utilization protein LamB\n",
            "\n",
            "fig|2116814.3352.peg.17\t14_13502_14197\tFIG015373: Membrane protein\n",
            "\n",
            "fig|2116814.3352.peg.18\t14_14199_15161\tFIG001614: Membrane protein\n",
            "\n",
            "fig|2116814.3352.peg.19\t14_15177_15824\tPyrrolidone-carboxylate peptidase (EC 3.4.19.3)\n",
            "\n",
            "fig|2116814.3352.peg.20\t14_15900_16625\tNAD-dependent protein deacetylase of SIR2 family\n",
            "\n",
            "fig|2116814.3352.peg.21\t14_16798_16926\thypothetical protein\n",
            "\n",
            "fig|2116814.3352.peg.22\t14_16981_19152\tPenicillin-binding protein\n",
            "\n",
            "fig|2116814.3352.peg.23\t14_19221_19748\tSignal peptidase I (EC 3.4.21.89)\n",
            "\n",
            "fig|2116814.3352.peg.24\t14_20295_19783\tCephalosporin hydroxylase\n",
            "\n",
            "fig|2116814.3352.peg.25\t14_20617_20480\thypothetical protein\n",
            "\n",
            "fig|2116814.3352.peg.26\t14_20574_21116\tMutT/nudix family protein\n",
            "\n",
            "fig|2116814.3352.peg.27\t14_22055_21294\tMethyltransferase (EC 2.1.1.-)\n",
            "\n",
            "fig|2116814.3352.peg.28\t14_22253_22744\tFIG01227518: hypothetical protein\n",
            "\n",
            "fig|2116814.3352.peg.29\t14_23595_22825\tDNA alkylation repair enzyme\n",
            "\n",
            "fig|2116814.3352.peg.30\t14_23934_23629\tFIG01226264: hypothetical protein\n",
            "\n",
            "fig|2116814.3352.peg.31\t14_25513_24047\tLysine-specific permease\n",
            "\n",
            "fig|2116814.3352.peg.32\t14_25737_25615\thypothetical protein\n",
            "\n",
            "fig|2116814.3352.peg.33\t14_26205_26654\tacetyltransferase, GNAT family\n",
            "\n",
            "fig|2116814.3352.peg.34\t14_27714_26695\tmembrane protein, putative\n",
            "\n",
            "fig|2116814.3352.peg.35\t14_27805_28605\tFIG01225599: hypothetical protein\n",
            "\n",
            "fig|2116814.3352.peg.36\t14_28616_29554\tAminoglycoside phosphotransferase\n",
            "\n",
            "fig|2116814.3352.peg.37\t14_29568_30311\tSerine/threonine protein phosphatase (EC 3.1.3.16)\n",
            "\n",
            "fig|2116814.3352.peg.38\t14_30435_31106\tUridine kinase (EC 2.7.1.48)\n",
            "\n",
            "fig|2116814.3352.peg.39\t14_31236_32060\tFIG01226434: hypothetical protein\n",
            "\n",
            "fig|2116814.3352.peg.40\t14_32166_33152\tFIG01229779: hypothetical protein\n",
            "\n",
            "fig|2116814.3352.peg.41\t14_33189_33860\tTwo-component response regulator ybdJ\n",
            "\n",
            "fig|2116814.3352.peg.42\t14_34019_34852\tsensor histidine kinase, putative\n",
            "\n",
            "fig|2116814.3352.peg.43\t14_34915_35229\tFIG01229360: hypothetical protein\n",
            "\n",
            "fig|2116814.3352.peg.44\t14_35756_35316\tEMG2 protein\n",
            "\n",
            "fig|2116814.3352.peg.45\t14_36135_36728\tFIG01227809: hypothetical protein\n",
            "\n",
            "fig|2116814.3352.peg.46\t14_38474_37032\tTranscriptional regulator, GntR family domain / Aspartate aminotransferase (EC 2.6.1.1)\n",
            "\n",
            "fig|2116814.3352.peg.47\t14_38599_39510\tPermease of the drug/metabolite transporter (DMT) superfamily\n",
            "\n",
            "fig|2116814.3352.peg.48\t14_39996_39529\tmutT/nudix family protein\n",
            "\n",
            "fig|2116814.3352.peg.49\t14_40093_40494\tFIG01227096: hypothetical protein\n",
            "\n",
            "fig|2116814.3352.peg.50\t14_40512_40982\tacetyltransferase, GNAT family\n",
            "\n",
            "fig|2116814.3352.peg.51\t14_41030_41920\thypothetical protein\n",
            "\n",
            "fig|2116814.3352.peg.52\t14_41941_42885\t6-aminohexanoate-dimer hydrolase (EC 3.5.1.46)\n",
            "\n",
            "fig|2116814.3352.peg.53\t14_43704_43069\tCAAX amino terminal protease family family\n",
            "\n",
            "fig|2116814.3352.peg.54\t14_43994_45019\tconserved domain protein\n",
            "\n",
            "fig|2116814.3352.peg.55\t14_45050_45586\tFIG01232871: hypothetical protein\n",
            "\n",
            "fig|2116814.3352.peg.56\t14_45618_46049\tFIG01231004: hypothetical protein\n",
            "\n",
            "fig|2116814.3352.peg.57\t14_46902_46111\tFIG01225478: hypothetical protein\n",
            "\n",
            "fig|2116814.3352.peg.58\t14_47099_47566\tNudix dNTPase DR1776 (EC 3.6.1.-)\n",
            "\n",
            "fig|2116814.3352.peg.59\t14_47587_48231\tFIG01227107: hypothetical protein\n",
            "\n",
            "fig|2116814.3352.peg.60\t14_50222_48279\tRibosome protection-type tetracycline resistance related proteins\n",
            "\n",
            "fig|2116814.3352.peg.61\t14_50487_50915\tTranscriptional regulator, MarR family\n",
            "\n",
            "fig|2116814.3352.peg.62\t14_50934_51572\tOxygen-insensitive NAD(P)H nitroreductase (EC 1.-.-.-) / Dihydropteridine reductase (EC 1.5.1.34)\n",
            "\n",
            "fig|2116814.3352.peg.63\t14_51619_51912\tFIG01225137: hypothetical protein\n",
            "\n",
            "fig|2116814.3352.peg.64\t14_52126_52401\tFIG01231425: hypothetical protein\n",
            "\n",
            "fig|2116814.3352.peg.65\t14_52442_52804\tFIG01230914: hypothetical protein\n",
            "\n",
            "fig|2116814.3352.peg.66\t14_52801_54435\tLmo0066 homolog within ESAT-6 gene cluster, similarity to ADP-ribosylating toxins\n",
            "\n",
            "fig|2116814.3352.peg.67\t14_54463_55710\tFIG01230945: hypothetical protein\n",
            "\n",
            "fig|2116814.3352.peg.68\t14_56012_56944\tATPase, AAA family\n",
            "\n",
            "fig|2116814.3352.peg.69\t14_56941_58050\tFIG01226901: hypothetical protein\n",
            "\n",
            "fig|2116814.3352.peg.70\t14_58066_58797\tFIG01225647: hypothetical protein\n",
            "\n",
            "fig|2116814.3352.peg.71\t14_59606_58962\tFIG01250685: hypothetical protein\n",
            "\n",
            "fig|2116814.3352.peg.72\t14_59745_60911\tD-alanyl-D-alanine carboxypeptidase (EC 3.4.16.4)\n",
            "\n",
            "fig|2116814.3352.peg.73\t14_60964_61119\tFIG01249153: hypothetical protein\n",
            "\n",
            "fig|2116814.3352.peg.74\t14_61268_63022\tAdenine deaminase (EC 3.5.4.2)\n",
            "\n",
            "fig|2116814.3352.peg.75\t14_63136_63648\tFIG01226544: hypothetical protein\n",
            "\n",
            "fig|2116814.3352.peg.76\t14_64152_66788\tMicrobial collagenase (EC 3.4.24.3)\n",
            "\n",
            "fig|2116814.3352.peg.77\t14_68391_67012\tCatalase (EC 1.11.1.6)\n",
            "\n",
            "fig|2116814.3352.peg.78\t14_68681_69898\tSuccinylornithine transaminase, putative\n",
            "\n",
            "46\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "{'6-aminohexanoate-dimer hydrolase (EC 3.5.1.46)',\n",
              " 'ATPase, AAA family',\n",
              " 'Adenine deaminase (EC 3.5.4.2)',\n",
              " 'Allophanate hydrolase 2 subunit 1 (EC 3.5.1.54)',\n",
              " 'Allophanate hydrolase 2 subunit 2 (EC 3.5.1.54)',\n",
              " 'Aminoglycoside phosphotransferase',\n",
              " 'Butyryl-CoA dehydrogenase (EC 1.3.99.2)',\n",
              " 'CAAX amino terminal protease family family',\n",
              " 'Catalase (EC 1.11.1.6)',\n",
              " 'Cephalosporin hydroxylase',\n",
              " 'Cytoplasmic copper homeostasis protein cutC',\n",
              " 'D-alanyl-D-alanine carboxypeptidase (EC 3.4.16.4)',\n",
              " 'DNA alkylation repair enzyme',\n",
              " 'EMG2 protein',\n",
              " 'EpiH/GdmH-related protein',\n",
              " 'FIG001614: Membrane protein',\n",
              " 'FIG015373: Membrane protein',\n",
              " 'Glyoxalase family protein',\n",
              " 'Lactam utilization protein LamB',\n",
              " 'Lmo0066 homolog within ESAT-6 gene cluster, similarity to ADP-ribosylating toxins',\n",
              " 'Lysine-specific permease',\n",
              " 'Methyltransferase (EC 2.1.1.-)',\n",
              " 'Microbial collagenase (EC 3.4.24.3)',\n",
              " 'MutT/nudix family protein',\n",
              " 'NAD-dependent protein deacetylase of SIR2 family',\n",
              " 'Nudix dNTPase DR1776 (EC 3.6.1.-)',\n",
              " 'Oxygen-insensitive NAD(P)H nitroreductase (EC 1.-.-.-) / Dihydropteridine reductase (EC 1.5.1.34)',\n",
              " 'Penicillin-binding protein',\n",
              " 'Permease of the drug/metabolite transporter (DMT) superfamily',\n",
              " 'Predicted histidine uptake transporter',\n",
              " 'Pyrrolidone-carboxylate peptidase (EC 3.4.19.3)',\n",
              " 'Ribosome protection-type tetracycline resistance related proteins',\n",
              " 'Serine/threonine protein phosphatase (EC 3.1.3.16)',\n",
              " 'Signal peptidase I (EC 3.4.21.89)',\n",
              " 'Succinylornithine transaminase, putative',\n",
              " 'Transcriptional regulator, ArsR family',\n",
              " 'Transcriptional regulator, GntR family domain / Aspartate aminotransferase (EC 2.6.1.1)',\n",
              " 'Transcriptional regulator, IclR family',\n",
              " 'Transcriptional regulator, MarR family',\n",
              " 'Two-component response regulator ybdJ',\n",
              " 'Uridine kinase (EC 2.7.1.48)',\n",
              " 'acetyltransferase, GNAT family',\n",
              " 'conserved domain protein',\n",
              " 'membrane protein, putative',\n",
              " 'mutT/nudix family protein',\n",
              " 'sensor histidine kinase, putative'}"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 60
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "olG2x31VXoCY",
        "outputId": "5f38bd00-d2e9-4e45-9459-78df87c75e3a"
      },
      "source": [
        "all_coordinates.shape"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "(60, 2)"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 54
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "jOK88nH2Y6qK"
      },
      "source": [
        "Статистика по RAST (далее в анализ не входили hypothetical proteins): точные функции установлены для 46 генов\n",
        "Статистика по PROKKA (далее в анализ не входили hypothetical proteins):точные функции установлены для 33 генов\n",
        "Статистика по аннотации: проаннотировано 60 участков \n",
        "\n",
        "Сравнение между prokka и аннотацией: 25 из 33 генов из prokka совпали по своей функции с аннотацией. То есть 8 функций генов нашлись в prokka (соответственно в моей аннотации 60-25 = 35)\n",
        "\n",
        "Сравнение между rast и аннотацией: совпадения по точным функциям - 7 из 46 (по rast). На самом деле данный анализ проводить сложнее, так как rast не выдавал gene name primary и тогда приходилось смотреть и сравнивать сначала с prokka (вдруг там есть совпадение и оттуда можно сразу получить gene name primary, а это упрощает поиск). Но оказалось, что и RAST с PROKKA между собой не сильно похожи (либо запись функций в rast настолько \"завуалированы\", что нужно вдаваться в биологию). Соответсвенно,  39 функций нашел rast в отличие от моей аннотации и 53 функции есть в моей аннотации, но нет в rast\n",
        " "
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "eBzX7WIWdkcl"
      },
      "source": [
        ""
      ],
      "execution_count": null,
      "outputs": []
    }
  ]
}